src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.EnumerateSplits.cs

```diff
@@ -0,0 +1,276 @@
+        /// </summary>
+        [StructLayout(LayoutKind.Auto)]
+        public ref struct ValueSplitEnumerator
+        {
+            private readonly Regex _regex;
+            private readonly ReadOnlySpan<char> _input;
+
+            private int _startAt;
+            private (int Index, int Length) _lastMatch;
+            private Range _currentSplit;
+            private int _remainingCount;
+
+            /// <summary>
+            /// Creates an instance of the <see cref="ValueSplitEnumerator"/> for the passed in <paramref name="regex"/> which iterates over <paramref name="input"/>.
+            /// </summary>
+            /// <param name="regex">The <see cref="Regex"/> to use for finding matches.</param>
+            /// <param name="input">The input span to iterate over.</param>
+            /// <param name="count">The maximum number of times the split can occur.</param>
+            /// <param name="startAt">The position from where the engine should start looking for matches.</param>
+            /// <param name="rtl">Whether the engine is matching from right to left.</param>
+            internal ValueSplitEnumerator(Regex regex, ReadOnlySpan<char> input, int count, int startAt, bool rtl)
+            {
+                _regex = regex;
+                _input = input;
+                _startAt = startAt;
+                _lastMatch = (rtl ? input.Length : 0, -1);
+                _remainingCount = count != 0 ? count : int.MaxValue; // Maintain same behavior as Split(..., count: 0, ...), which treats it as effectively infinite.
+            }
+
+            /// <summary>Provides an enumerator that iterates through the splits in the input span.</summary>
+            /// <returns>A copy of this enumerator.</returns>
+            public readonly ValueSplitEnumerator GetEnumerator() => this;
+
+            /// <summary>Advances the enumerator to the next split.</summary>
+            /// <returns>
+            /// <see langword="true"/> if the enumerator was successfully advanced to the next element; <see langword="false"/> if the enumerator cannot find additional matches.
+            /// </returns>
+            public bool MoveNext()
+            {
+                // If we've already found all the splits, we're done.
+                if (_remainingCount == 0)
+                {
+                    return false;
+                }
+
+                if (_remainingCount == 1)
+                {
+                    // If we've reached the last split, include everything that remains.
+                    _currentSplit = !_regex.RightToLeft ? (_lastMatch.Index + _lastMatch.Length).._input.Length : 0.._lastMatch.Index;
+                }
+                else
+                {
+                    // Perform the next match.
+                    (bool Success, int Index, int Length, int TextPosition) match = _regex.RunSingleMatch(RegexRunnerMode.BoundsRequired, _lastMatch.Length, _input, _startAt);
```

Comment:
I forget, Would this also optimize away capture groups given we don't need them? I was sort of assuming that the implementation here would use EnumerateMatches internally and then just produce the right range from that result.

Reply:
> Would this also optimize away capture groups given we don't need them?

Yes, or at least it will for non-backtracking, where BoundsRequired avoids the need to do the last pass that handles generating capture information.

> I was sort of assuming that the implementation here would use EnumerateMatches internally and then just produce the right range from that result.

I had that initially, and then hit a snag that caused me to move away from it, but I can't recall what that was now.